File: backend/src/main/kotlin/app/meals/service/ShoppingListService.kt
package app.meals.service

import app.meals.domain.DayAssignment
import app.meals.domain.MealPlanDoc
import app.meals.domain.RecipeDoc
import app.meals.domain.ShoppingListItem
import app.meals.domain.UnitFamily
import app.meals.domain.bestDisplayUnit
import app.meals.domain.normalizeUnit
import java.util.UUID
import kotlin.math.ceil

private sealed class ShoppingContribution {
    data class InFamily(val baseAmount: Double, val recipeId: String) : ShoppingContribution()

    data class Raw(val quantity: String, val recipeId: String) : ShoppingContribution()
}

object ShoppingListService {

    /**
     * Aggregates ingredients across assignments; merges by known unit family (volume/weight) when quantity is numeric.
     */
    fun buildAggregatedShoppingItems(
        plan: MealPlanDoc,
        recipeById: Map<String, Pair<UUID, RecipeDoc>>,
    ): List<ShoppingListItem> {
        val aggregated = mutableMapOf<String, MutableList<ShoppingContribution>>()
        for (assignment in plan.assignments) {
            val recipeId = assignment.recipeId
            val (recipeUuid, doc) = recipeById[recipeId] ?: continue
            val scale = scaleForAssignment(plan, assignment, doc)
            val idStr = recipeUuid.toString()
            for (ing in doc.ingredients) {
                val name = ing.name.lowercase().trim()
                val known = normalizeUnit(ing.unit)
                val scaledQtyStr = scaledIngredientQuantity(ing.quantity, scale)
                val parsed = parseQuantity(scaledQtyStr)
                val key =
                    if (known != null && parsed != null) {
                        "$name|${known.family.name}"
                    } else {
                        "$name|${ing.unit}"
                    }
                val list = aggregated.getOrPut(key) { mutableListOf() }
                if (known != null && parsed != null) {
                    list.add(ShoppingContribution.InFamily(parsed * known.toBase, idStr))
                } else {
                    list.add(ShoppingContribution.Raw(scaledQtyStr, idStr))
                }
            }
        }
        return aggregated.map { (key, contribs) ->
            val (nameLower, second) = key.split("|", limit = 2).let { parts ->
                parts[0] to parts.getOrElse(1) { "" }
            }
            val displayName = nameLower.replaceFirstChar { it.uppercase() }
            val recipeIdsUsed = contribs.map { c ->
                when (c) {
                    is ShoppingContribution.InFamily -> c.recipeId
                    is ShoppingContribution.Raw -> c.recipeId
                }
            }.distinct()
            when (second) {
                UnitFamily.VOLUME.name, UnitFamily.WEIGHT.name -> {
                    val family = UnitFamily.valueOf(second)
                    val sum = contribs.filterIsInstance<ShoppingContribution.InFamily>().sumOf { it.baseAmount }
                    val (displayVal, unit) = bestDisplayUnit(sum, family)
                    ShoppingListItem(
                        name = displayName,
                        quantity = formatQuantity(displayVal),
                        unit = unit,
                        recipeIds = recipeIdsUsed,
                    )
                }
                else -> {
                    val quantities =
                        contribs.filterIsInstance<ShoppingContribution.Raw>().map { it.quantity }.filter { it.isNotBlank() }
                    val combinedQty = summarizeQuantities(quantities)
                    ShoppingListItem(
                        name = displayName,
                        quantity = combinedQty,
                        unit = second,
                        recipeIds = recipeIdsUsed,
                    )
                }
            }
        }.sortedBy { it.name }
    }

    fun scaleForAssignment(plan: MealPlanDoc, assignment: DayAssignment, doc: RecipeDoc): Double {
        val effective = assignment.persons ?: plan.defaultPersons
        if (effective == null) return 1.0
        val servings = doc.servings.coerceAtLeast(1)
        return effective.toDouble() / servings
    }

    /**
     * Scales a recipe line quantity for shopping list aggregation.
     * Numeric strings are multiplied by [scale]; non-numeric use [ceil] of scale as a discrete count prefix (e.g. `2× pinch`).
     */
    fun scaledIngredientQuantity(rawQuantity: String, scale: Double): String {
        val q = rawQuantity.trim()
        if (q.isBlank()) return q
        val parsed = parseQuantity(q)
        return if (parsed != null) {
            formatQuantity(parsed * scale)
        } else {
            val n = ceil(scale).toInt().coerceAtLeast(1)
            if (n == 1) q else "${n}× $q"
        }
    }

    /**
     * Summarizes a list of quantity strings: if all parse as numbers, returns their sum (e.g. "200" + "300" -> "500");
     * otherwise returns them joined with ", " (e.g. "pinch", "handful" -> "pinch, handful").
     */
    fun summarizeQuantities(quantities: List<String>): String {
        if (quantities.isEmpty()) return "—"
        val parsed = quantities.mapNotNull { q -> parseQuantity(q) }
        return if (parsed.size == quantities.size) {
            val sum = parsed.sum()
            formatQuantity(sum)
        } else {
            quantities.distinct().joinToString(", ")
        }
    }
}

private fun parseQuantity(s: String): Double? = s.trim().replace(",", ".").toDoubleOrNull()

private fun formatQuantity(value: Double): String =
    if (value == value.toLong().toDouble()) "${value.toLong()}" else "%.2f".format(value).trimEnd('0').trimEnd('.')

File: backend/src/service/shoppingListService.ts
import { boughtSourceKey, shoppingItemKey } from "../domain/shoppingItemKey.js";
import type {
  DayAssignment,
  ItemSource,
  MealPlanDoc,
  RecipeDoc,
  ShoppingListItem,
} from "../domain/types.js";
import { bestDisplayUnit, normalizeUnit, type UnitFamily } from "../domain/unitConversion.js";

export interface RecipeEntry {
  id: string;
  doc: RecipeDoc;
}

type Contribution =
  | { kind: "inFamily"; baseAmount: number; source: ItemSource }
  | { kind: "raw"; quantity: string; source: ItemSource };

export interface AggregateOptions {
  /**
   * Contributions already bought on a completed trip, as `boughtSourceKey`s
   * (see `domain/shoppingItemKey.ts`). They are left out of the aggregate so
   * the open list only shows what is still to buy.
   */
  bought?: ReadonlySet<string>;
}

/**
 * Aggregate ingredients across meal-plan assignments. Same-name ingredients merge
 * by unit family (volume/weight) when the quantity is numeric; otherwise they stay
 * grouped by their raw unit. Ported from the Kotlin ShoppingListService.
 */
export function buildAggregatedShoppingItems(
  plan: MealPlanDoc,
  recipeById: Map<string, RecipeEntry>,
  options: AggregateOptions = {},
): ShoppingListItem[] {
  const aggregated = new Map<string, Contribution[]>();
  const bought = options.bought;

  for (const assignment of plan.assignments) {
    const entry = recipeById.get(assignment.recipeId);
    if (!entry) continue;
    const scale = scaleForAssignment(plan, assignment, entry.doc);
    const source: ItemSource = { day: assignment.day, recipeId: entry.id };

    for (const ing of entry.doc.ingredients) {
      const name = ing.name.toLowerCase().trim();
      const known = normalizeUnit(ing.unit);
      const scaledQty = scaledIngredientQuantity(ing.quantity, scale);
      const parsed = parseQuantity(scaledQty);
      const key = known && parsed !== null ? `${name}|${known.family}` : `${name}|${ing.unit}`;

      if (
        bought?.has(
          boughtSourceKey(shoppingItemKey({ name: ing.name, quantity: scaledQty, unit: ing.unit }), source),
        )
      ) {
        continue;
      }

      let list = aggregated.get(key);
      if (!list) {
        list = [];
        aggregated.set(key, list);
      }
      if (known && parsed !== null) {
        list.push({ kind: "inFamily", baseAmount: parsed * known.toBase, source });
      } else {
        list.push({ kind: "raw", quantity: scaledQty, source });
      }
    }
  }

  const items: ShoppingListItem[] = [];
  for (const [key, contribs] of aggregated) {
    const sep = key.indexOf("|");
    const nameLower = key.slice(0, sep);
    const second = key.slice(sep + 1);
    const displayName = nameLower.charAt(0).toUpperCase() + nameLower.slice(1);
    const recipeIds = [...new Set(contribs.map((c) => c.source.recipeId))];
    const sources = distinctSources(contribs.map((c) => c.source));

    if (second === "volume" || second === "weight") {
      const sum = contribs.reduce((acc, c) => (c.kind === "inFamily" ? acc + c.baseAmount : acc), 0);
      const [displayVal, unit] = bestDisplayUnit(sum, second as UnitFamily);
      items.push({ name: displayName, quantity: formatQuantity(displayVal), unit, recipeIds, sources });
    } else {
      const quantities = contribs
        .filter((c): c is Extract<Contribution, { kind: "raw" }> => c.kind === "raw")
        .map((c) => c.quantity)
        .filter((q) => q.trim() !== "");
      items.push({
        name: displayName,
        quantity: summarizeQuantities(quantities),
        unit: second,
        recipeIds,
        sources,
      });
    }
  }
  return items.sort((a, b) => (a.name < b.name ? -1 : a.name > b.name ? 1 : 0));
}

function distinctSources(sources: ItemSource[]): ItemSource[] {
  const byKey = new Map<string, ItemSource>();
  for (const s of sources) byKey.set(`${s.day}\u0000${s.recipeId}`, s);
  return [...byKey.values()];
}

/**
 * Scale factor for an assignment: effective persons ÷ recipe servings (default 1).
 * Falls back to 1 whenever any input is missing or non-finite so a bad persons
 * or servings value can never poison a quantity with NaN (which would drop an
 * ingredient out of its unit family and split it into a second "NaN" row).
 */
export function scaleForAssignment(
  plan: MealPlanDoc,
  assignment: DayAssignment,
  doc: RecipeDoc,
): number {
  const effective = assignment.persons ?? plan.defaultPersons;
  if (effective == null || !Number.isFinite(effective)) return 1;
  const servings = Math.max(Number.isFinite(doc.servings) ? doc.servings : 1, 1);
  const scale = effective / servings;
  return Number.isFinite(scale) && scale > 0 ? scale : 1;
}

/**
 * Scale a recipe line quantity. Numeric strings are multiplied; non-numeric ones
 * use ceil(scale) as a discrete count prefix (e.g. "2× pinch").
 */
export function scaledIngredientQuantity(rawQuantity: string, scale: number): string {
  const q = rawQuantity.trim();
  if (!q) return q;
  const parsed = parseQuantity(q);
  if (parsed !== null) return formatQuantity(parsed * scale);
  const n = Math.max(Math.ceil(scale), 1);
  return n === 1 ? q : `${n}× ${q}`;
}

/**
 * Sum quantities when all parse as numbers; otherwise join distinct strings with
 * ", ". Returns "—" for an empty list.
 */
export function summarizeQuantities(quantities: string[]): string {
  if (quantities.length === 0) return "—";
  const parsed = quantities.map(parseQuantity).filter((n): n is number => n !== null);
  if (parsed.length === quantities.length) {
    return formatQuantity(parsed.reduce((a, b) => a + b, 0));
  }
  return [...new Set(quantities)].join(", ");
}

function parseQuantity(s: string): number | null {
  const cleaned = s.trim().replace(/,/g, ".");
  if (cleaned === "") return null;
  const n = Number(cleaned);
  return Number.isFinite(n) ? n : null;
}

function formatQuantity(value: number): string {
  if (value === Math.trunc(value)) return String(Math.trunc(value));
  return value.toFixed(2).replace(/0+$/, "").replace(/\.$/, "");
}
